crates/ironrdp-cliprdr-format/src/bitmap.rs

```diff
@@ -481,62 +482,56 @@ struct PngEncoderInput {
 ///
 /// INVARIANT: bit_count <= 32
 #[allow(clippy::arithmetic_side_effects)]
-fn bmp_stride(width: u16, bit_count: u16) -> usize {
+fn rgb_bmp_stride(width: u16, bit_count: u16) -> usize {
     debug_assert!(bit_count <= 32);
     (((usize::from(width) * usize::from(bit_count)) + 31) & !31) >> 3
 }
 
-fn transform_bitmap(
+fn bgra_to_top_down_rgba(
     header: &BitmapInfoHeader,
-    input: &[u8],
+    src_bitmap: &[u8],
     preserve_alpha: bool,
-) -> Result<PngEncoderInput, BitmapError> {
-    // If height is positive, DIB is bottom-up, but target PNG format is top-down.
-    let flip = header.flip_vertically();
+) -> Result<PngEncoderContext, BitmapError> {
+    // DIB may be encoded bottom-up, but the format we target, PNG, is top-down.
+    let should_flip_vertically = header.is_bottom_up();
 
     let width = header.width();
     let height = header.height();
 
-    let bit_count = header.bit_count;
+    let src_n_samples = usize::from(header.bit_count / 8);
 
-    let stride = bmp_stride(width, bit_count);
+    let src_stride = rgb_bmp_stride(width, header.bit_count);
 
-    let input_bytes_per_pixel = usize::from(bit_count / 8);
-    let color_type = if preserve_alpha {
-        png::ColorType::Rgba
+    let (dst_color_type, dst_n_samples) = if preserve_alpha {
+        (png::ColorType::Rgba, 4)
     } else {
-        png::ColorType::Rgb
+        (png::ColorType::Rgb, 3)
     };
 
-    let components = color_type.samples();
-    debug_assert!(components <= 4);
-
     // INVARIANT: height * width * components <= u16::MAX * u16::MAX * 4 < usize::MAX
     // This is always true because `components <= 4` is checked above, and width & height
     // bounds are validated on PDU encode/decode
     #[allow(clippy::arithmetic_side_effects)]
-    let frame_buffer_len = usize::from(height) * usize::from(width) * components;
-
-    // Prevent allocation of huge frame buffers
-    check_invariant(frame_buffer_len <= MAX_BUFFER_SIZE).ok_or(BitmapError::BufferTooBig)?;
+    let dst_bitmap_len = usize::from(height) * usize::from(width) * dst_n_samples;
 
-    let mut frame_buffer = vec![0u8; frame_buffer_len];
+    // Prevent allocation of huge buffers.
+    ensure(dst_bitmap_len <= MAX_BUFFER_SIZE).ok_or(BitmapError::BufferTooBig)?;
 
-    let mut strides_normal;
-    let mut strides_reversed;
+    let mut rows_normal;
+    let mut rows_reversed;
 
-    let strides: &mut dyn Iterator<Item = &[u8]> = if flip {
-        strides_reversed = input.chunks_exact(stride).rev();
-        &mut strides_reversed
+    let rows: &mut dyn Iterator<Item = &[u8]> = if should_flip_vertically {
+        rows_reversed = src_bitmap.chunks_exact(src_stride).rev();
+        &mut rows_reversed
     } else {
-        strides_normal = input.chunks_exact(stride);
-        &mut strides_normal
+        rows_normal = src_bitmap.chunks_exact(src_stride);
+        &mut rows_normal
     };
```

Comment:
The stride of a bitmap is the _number of bytes_ for a pixel row (i.e.: a scan line). Given the stride of a bitmap, we can iterate over the rows, but we don’t iterate on "strides".